src/internal/operators/distinctUntilChanged.ts

```diff
@@ -1,78 +1,181 @@
+ * equality. Defaults to a `===` check.
+ */
+export function distinctUntilChanged<T>(comparator?: (previous: T, current: T) => boolean): MonoTypeOperatorFunction<T>;
+
+/**
+ * Returns a result {@link Observable} that emits all values pushed by the source observable if they
+ * are distinct in comparison to the last value the result observable emitted.
+ *
+ * 1. It will always emit the first value from the source.
+ * 2. The `keySelector` will be run against all values, including the first value.
+ * 3. For all values after the first, the selected key will be compared against the key selected from
+ *    the previously emitted value using the `comparator`.
+ * 4. If the keys are determined to be unequal by this check, the value (not the key), is emitted
+ *    and the selected key from that value is saved for future comparisons against other keys.
+ *
+ * ## Example
+ *
+ * Selecting update events only when the `updatedBy` field shows
+ * the account changed hands...
+ *
+ * ```ts
+ * // A stream of updates to a given account
+ * const accountUpdates$ = of(
+ *   { updatedBy: "blesh", data: [] },
+ *   { updatedBy: "blesh", data: [] },
+ *   { updatedBy: "ncjamieson", data: [] },
+ *   { updatedBy: "ncjamieson", data: [] },
+ *   { updatedBy: "blesh", data: [] }
+ * );
+ *
+ * // We only want the events where it changed hands
+ * const changedHands$ = accountUpdates$.pipe(
+ *   distinctUntilChanged(undefined, update => update.updatedBy)
+ * );
  *
- * @param {function} [compare] Optional comparison function called to test if an item is distinct from the previous item in the source.
- * A return value of true indicates that it is the same, and a return value of false means they are different.
- * @return {Observable} An Observable that emits items from the source Observable with distinct values.
+ * changedHands$.subscribe(console.log);
+ * // Logs:
+ * // {updatedBy: "blesh", data: Array[0]}
+ * // {updatedBy: "ncjamieson", data: Array[0]}
+ * // {updatedBy: "blesh", data: Array[0]}
+ * ```
+ *
+ * @param comparator A function used to compare the previous and current keys for
+ * equality. Defaults to a `===` check.
+ * @param keySelector Used to select a key value to be passed to the `comparator`.
  */
-export function distinctUntilChanged<T, K>(compare?: (a: K, b: K) => boolean, keySelector?: (x: T) => K): MonoTypeOperatorFunction<T> {
-  compare = compare ?? defaultCompare;
+export function distinctUntilChanged<T, K>(
+  comparator: (previous: K, current: K) => boolean,
+  keySelector: (value: T) => K
+): MonoTypeOperatorFunction<T>;
+
+export function distinctUntilChanged<T, K>(
+  comparator?: (previous: K, current: K) => boolean,
+  keySelector: (value: T) => K = identity as (value: T) => K
+): MonoTypeOperatorFunction<T> {
+  // We've been allowing `null` do be passed as the `compare`, so we can't do
+  // a default value for the parameter, because that will only work
+  // for `undefined`.
+  comparator = comparator ?? defaultCompare;
```

Comment:
Added this comment, because I tried to use a default value parameter here and it was sad. We should probably deprecate this sort of thing at some point, and make this a configuration object, I suppose.